### ISSUE TITLE
Dynamical Decomposition: LOSVD Inconsistencies

### ISSUE THREAD

**Issue by adriano-poci** [dynamics-of-stellar-systems/dynamite#229]
The `modelgen` binary allows for the decomposition of a given Schwarzschild model into many dynamical components. `modelgen` re-integrates the original model in sub-sections in order to separate dynamical components.

In principle, the combination of all dynamical components should be equivalent to the original full Schwarzschild model. However, there is currently a discrepancy between the combined LOSVD over all dynamical components, and the original model they were derived from, in any given aperture.

## Behaviour of Issue
The discrepancy appears to increase with increasing number of dynamical components. There still exists a small discrepancy even for a single-component re-integration.

![losvdComparison](https://user-images.githubusercontent.com/19634038/171187577-7c8b58e6-dd32-4bbf-a49c-1d8d1bd865c0.png)
Example of the discrepancy. Left panel shows the single-component re-integration. Right shows a large number of dynamical components (144). In both panels, the grey filled histogram is from the original Schwarzschild, the black solid line is the combined LOSVD over all dynamical components, and the coloured histograms are individual components.

## Possible Causes
* In the config file passed to the `modelgen` binary (here `modelgen.in`), there are two arguments representing "accuracies" of the orbit integration.
* Possible impact of new random seed for each orbit integration.

## Attempted Solutions
* So far, I have only attempted different "re-integration" accuracies (second-last parameter in `modelgen.in`). Even down to precisions of `~ 1e-8`, which results in prohibitively long integrations times, the discrepancy remains.

## Things to Try
* Systematically varying the accuracy parameters, even for a single component, to test if the discrepancy can be minimised/removed.
* Test if fixing the random seed affects the re-integration.

**Comment by prashjet** [comment]
It's super suggestive that the spike at ~600 km/s appears at the same position on the left and right, no....? If it is the same position, I think it's likely that those orbits have been re-integrated faithfully, and maybe the discrepancy is a normalisation/re-weighting issue...

@adriano-poci @gsantucci90 to recombine them, do you need to re-weight the orbits produced by modelgen?

**Comment by sthater** [comment]
Notes on the conversation with the MAGPI team:

* random seed is discussed again
* do a test with BayesLOSVD (?)

**Comment by adriano-poci** [comment]
@prashjet It's not re-weighting the orbits, but rather re-normalising the histograms. The histograms of the separate dynamical components are re-normalised to conserve the total mass in a given aperture. That is, the integral of all sub-component LOSVD should be equal to the integral of the original (full-model) LOSVD in that aperture.
This doesn't change the shape of individual LOSVD, but only affects the shape of the combined total LOSVD in a given aperture.

But I combed over this part of the code thoroughly to make sure it was doing the normalisation correctly. And I think @gsantucci90 independently discovered the same issue using independent analysis scripts. Moreover, the left and right figures are re-normalised using the same code.

What _might_ be happening is that there is no relative weighting between the different components when they come out of `modelgen`. Currently the re-normalisation will preserve the relative integrals of all the sub-components, in every aperture. But if this relative weighting is wiped out during `modelgen`, then we would have to somehow manually re-weight every dynamical component in every aperture (which isn't the same summing the orbital weights of all the orbits which contribute to that component).

**Comment by adriano-poci** [comment]
Progress update:
I have switched to reading in the full orbit library binary file, rather than re-integrating orbits with `modelgen`. I do this using a wrapped version of the `Histogram` class from `dynamite`. After reading in the full histograms, they are then convolved by the best-fitting orbital weights, and scaled by the global "gravitational" scale factor. I then apply any sort of dynamical selection, as previously, decompose the components, then re-combine them to test the resulting summed LOSVD.

Initial tests showed similar problems to above - seemingly random mis-matches between the summed LOSVD and the original LOSVD in any given aperture.

After some experimentation, it seems that in this particular case, `Histogram.normalise` slightly changes the shape of the summed histogram. I haven't looked into how this is possible yet.
But omitting `Histogram.normalise` (and manually renormalising the components), I get the following results for random apertures:
[LOSVD_ii_4_10.pdf](https://github.com/dynamics-of-stellar-systems/dynamite/files/11102367/LOSVD_ii_4_10.pdf)
[LOSVD_ii_4_750.pdf](https://github.com/dynamics-of-stellar-systems/dynamite/files/11102421/LOSVD_ii_4_750.pdf)



So it seems that this is working as expected now. The things to check are:

- What is happening with `Histogram.normalise`
- The current `orbit_exploration` ignores the orbital weights. The orbit library is integrated without knowledge of the kinematics, and so it needs to be convolved with the orbital weights in order to produce the actual best-fit set of LOSVD. Somewhere the code needs a
`losvd_histograms.y = np.multiply(losvd_histograms.y, weights[:, np.newaxis, np.newaxis])`
assuming axis 0 of `losvd_histograms` is `N_Orbs`

**Comment by maindlt** [comment]
Hi Adriano,

> * The current `orbit_exploration` ignores the orbital weights. The orbit library is integrated without knowledge of the kinematics, and so it needs to be convolved with the orbital weights in order to produce the actual best-fit set of LOSVD. Somewhere the code needs a
>   `losvd_histograms.y = np.multiply(losvd_histograms.y, weights[:, np.newaxis, np.newaxis])`
>   assuming axis 0 of `losvd_histograms` is `N_Orbs`

Actually, `Decomposition.create_orbital_component_files_giu()` in the current `orbit_exploration.py` does write orbital weights into temporary files `<component>_orb_s22.out` which are subsequently read by `Decomposition.comps_aphist()` and convolved with the losvds (line 290 in the current master branch).
Please note that there is a new implementation of the decomposition in the works which will include a major code cleanup including re-use of many DYNAMITE classes and methods and possibly a rename of the file `orbit_exploration.py`...

Cheers,
Thomas

**Comment by maindlt** [comment]
Hi @adriano-poci and @prashjet,

it's been a while and the main points of this Issue have been addressed:
1. The seemingly problematic `Histogram.normalise()` method is not used anywhere in the DYNAMITE code any more.
2. Convolution with the weights is done in the `Decomposition` class in `analysis.py`, more specifically in `Decomposition.comps_aphist()` lines 148/149:
```           
losvd = np.dot(self.losvd_histograms.y[orb_sel,:,:].T,
               self.model.weights[orb_sel]).T
```

Therefore:
- Can we close this issue?
- Shall we remove the `Histogram.normalise()` method altogether to avoid future confusion? This will also involve removing about a handful `normalise=False` statements in the code. If you vote 'yes', I'll create a PR for it.

Thanks and cheers, Thomas

**Comment by adriano-poci** [comment]
Agreed. I've just seen the new decomposition commit, so will have to check that out, but in any case it seems this issue can be closed.

Since it is deprecated, it makes sense to remove `Histogram.normalise`, I suppose.

Thanks.
